zebra-chain/src/commitments/sapling.rs

```diff
@@ -0,0 +1,341 @@
+//! Sapling note and value commitments and types.
+
+#![allow(clippy::unit_arg)]
+
+#[cfg(test)]
+mod arbitrary;
+#[cfg(test)]
+mod test_vectors;
+
+use std::{fmt, io};
+
+use bitvec::prelude::*;
+use rand_core::{CryptoRng, RngCore};
+
+use crate::{
+    keys::sapling::{find_group_hash, Diversifier, TransmissionKey},
+    serde_helpers,
+    serialization::{ReadZcashExt, SerializationError, ZcashDeserialize, ZcashSerialize},
+    types::amount::{Amount, NonNegative},
+};
+
+/// Generates a random scalar from the scalar field 𝔽_{r_𝕁}.
+///
+/// The prime order subgroup 𝕁^(r) is the order-r_𝕁 subgroup of 𝕁 that consists
+/// of the points whose order divides r. This function is useful when generating
+/// the uniform distribution on 𝔽_{r_𝕁} needed for Sapling commitment schemes'
+/// trapdoor generators.
+///
+/// https://zips.z.cash/protocol/protocol.pdf#jubjub
+pub fn generate_trapdoor<T>(csprng: &mut T) -> jubjub::Fr
+where
+    T: RngCore + CryptoRng,
+{
+    let mut bytes = [0u8; 64];
+    csprng.fill_bytes(&mut bytes);
+    // Fr::from_bytes_wide() reduces the input modulo r via Fr::from_u512()
+    jubjub::Fr::from_bytes_wide(&bytes)
+}
+
+/// "...an algebraic hash function with collision resistance (for fixed input
+/// length) derived from assumed hardness of the Discrete Logarithm Problem on
+/// the Jubjub curve."
+///
+/// PedersenHash is used in the definitions of Pedersen commitments (§
+/// 5.4.7.2‘Windowed Pedersen commitments’), and of the Pedersen hash for the
+/// Sapling incremental Merkle tree (§ 5.4.1.3 ‘MerkleCRH^Sapling Hash
+/// Function’).
+///
+/// https://zips.z.cash/protocol/protocol.pdf#concretepedersenhash
+#[allow(non_snake_case)]
+pub fn pedersen_hash_to_point(domain: [u8; 8], M: &BitVec<Lsb0, u8>) -> jubjub::ExtendedPoint {
+    // Expects i to be 0-indexed
```

Comment:
![image](https://user-images.githubusercontent.com/44879/89446546-838db080-d709-11ea-9147-e36ec32a7362.png)

Assuming this is the same `I_i`, it looks like there's a computation of `i - 1`, but here `i` is a `u32`, so it's not allowed to compute `0 - 1`.  When you write "expects i to be 0-indexed", do you mean that this `I_i` is defined differently to the spec so that it takes a "pre-decremented" value?

Reply:
Good point. I can make `I_i` expect a 1-indexed value, and then pass the 0-indexed bytes to `find_group_hash`

Reply:
(Although `N` in the spec is technically the non-negative integers, which includes zero, vs the positive integers N+, which doesn't include zero, but I get what they _mean_ here)